dips/dip-168.md

```diff
@@ -0,0 +1,124 @@
+2. **Concurrency:** Multiple transactions should be able to be in-flight at the same time, and can be processed in any order with a CRSN.
+3. **Force Expiration:** It must be hard to accidentally invalidate a given CRSN and sequence nonce pair without processing a transaction that this pair is associated with. E.g., you should only be able to expire a transaction knowingly, by sending another transaction with the same sequence number/nonce, or by sending a specific transaction to expire a transaction. A transaction should never expire as a result of another transaction with a different sequence number/nonce being processed with the exception of a force expiration transaction, as this makes it impossible for mempool to determine when to remove old transactions that can no longer be processed.
+4. **Dependency:** Dependency (but not strict sequentiality) must be able to be expressed between two transactions. e.g., an account should be able to guarantee that `txn2` will not be executed before `txn1` by some other means than waiting for `txn1` to be executed and committed before submitting `txn2`.
+5. **Space:** The representation of the CRSN must have a constant-size space footprint.
+6. **Opt-in:** Using a CRSN must be opt-in at the account level. An account must be able to opt-out of using a CRSN at any point. The system must be able to support accounts with normal sequence numbers and other accounts with CRSNs coexisting in the network.
+
+# CRSN Definition
+
+We define a CRSN as follows. A CRSN `CRSN(k)` is a stateful object published under an account `A` and parameterized by a bounded number `k`. `CRSN(k)` holds two pieces of data:
+
+* A `min_nonce: u64` field
+* A `slots: [bool; k]` field consisting of `k` boolean slots initialized to `false`.
+
+Each transaction sent from `A` is assigned a number or *sequence nonce* `i: u64`. This number coupled with the state `CRSN(k)` held under `A` defines the acceptance and rejection criteria for transactions.
+
+## Acceptance and Rejection Criteria
+
+The acceptance and rejection criteria for a pair `(i, CRSN(k))` where `i` is the sequence nonce assigned to a transaction sent by `A` and `CRSN(k)` is the CRSN state under `A` at the time of processing is as follows:
+
+1. If `i` is greater than or equal to `CRSN(k).min_nonce + k` the transaction    is discarded.
+2. If `i` is less than `CRSN(k).min_nonce` the transaction is discarded.
+3. If `i` is greater than or equal to `CRSN(k).min_nonce` and less than    `CRSN(k).min_nonce + k` then
+    1.  If the slot at position `i` in `CRSN(k).slots` is `false` then `CRSN(k).slots[i]` is set to `true` and the transaction is accepted. Subsequently, the [shifting algorithm](#shifting-algorithm) is applied to `CRSN(k)`. `A`'s sequence number is incremented.
+    2. If the slot at position `i` in `CRSN(k).slots` is `true` the transaction is rejected.
```

Comment:
don't rejects cost gas? is the assumption that a validator mempool would note this as already committed and discard it?

Reply:
Reject is gas free and can happen in a mempool or during execution.  Aborts are the transactions that pay gas and appear on-chain.

Reply:
ah, is it discard or reject, 🤦